activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, options)
+      @model = model
+      @inserts = inserts
+      @options = options
+    end
+
+    def execute
+      return ActiveRecord::Result.new([], []) if inserts.empty?
+
+      connection.exec_query to_sql, "Bulk Insert"
+    end
+
+    def to_sql
+      raise ArgumentError, "inserts is empty" if inserts.empty?
+
+      connection.build_insert_all_sql(
+        table_name,
+        keys,
+        values_list,
+        updatable_columns: updatable_columns,
+        returning: returning,
+        on_duplicate: on_duplicate,
+        conflict_target: conflict_target)
+    end
+
+    private
+      attr_reader :model, :inserts, :options
+
+      delegate :connection,
+               :table_name,
+               to: :model
+
+      delegate :supports_insert_returning?,
+               :supports_insert_conflict_target?,
+               :supports_insert_on_duplicate_skip?,
+               :supports_insert_on_duplicate_update?,
+               to: :connection
+
+      def primary_keys
+        Array.wrap(model.primary_key)
+      end
+
+      def readonly_columns
+        primary_keys + model.readonly_attributes.to_a
+      end
+
+      def keys
+        @keys ||= inserts.first.keys.map(&:to_s)
+      end
+
+      def values_list
+        return @values_list if defined?(@values_list)
+        _keys, @values_list = connection.prepare_keys_and_values_for_insert(inserts, table_name)
+        @values_list
+      end
+
+      def returning
+        return @returning if defined?(@returning)
+        @returning = options.fetch(:returning) { supports_insert_returning? && primary_keys }
+        @returning = false if @returning == []
+        if @returning && !supports_insert_returning?
+          raise ArgumentError, "#{connection} does not support :returning"
+        end
+        @returning
+      end
+
+      def conflict_target
+        return @conflict_target if defined?(@conflict_target)
+        @conflict_target = options[:conflict_target]
+        if @conflict_target && !supports_insert_conflict_target?
+          raise ArgumentError, "#{connection} does not support :conflict_target"
+        end
+        @conflict_target
+      end
+
+      def updatable_columns
+        return @updatable_columns if defined?(@updatable_columns)
+        @updatable_columns = keys - readonly_columns
+        @updatable_columns -= conflict_target.fetch(:columns).map(&:to_s) if conflict_target
+        @updatable_columns
+      end
+
+      def on_duplicate
+        return @on_duplicate if defined?(@on_duplicate)
```

Comment:
Why do we need ivar memoization at all here? Where is this called multiple times?

Also, need a case statement with clear exit points.